drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c

```diff
@@ -0,0 +1,426 @@
+		LOG_ERR("Failed Setting OTP voltage IOVDD to 2.5V\n");
+		return -ENOEXEC;
+	}
+	write_word(OTP_VOLTCTRL_ADDR, OTP_VOLTCTRL_2V5);
+	return 0;
+}
+
+static int poll_otp_read_valid(void)
+{
+	int otp_mem_status = 0;
+	int poll = 0;
+
+	while (poll < 100) {
+		read_word(OTP_POLL_ADDR, &otp_mem_status);
+
+		if ((otp_mem_status&0x2) == OTP_READ_VALID) {
+			return 0;
+		}
+		poll++;
+	}
+	LOG_ERR("failed poll_otp_read_valid()\n");
+	return -ENOEXEC;
+}
+
+static int poll_otp_wrdone(void)
+{
+	int otp_mem_status = 0;
+	int poll = 0;
+
+	while (poll < 100) {
+		read_word(OTP_POLL_ADDR, &otp_mem_status);
+
+		if ((otp_mem_status&0x1) == OTP_WR_DONE) {
+			return 0;
+		}
+		poll++;
+	}
+
+	LOG_ERR("failed poll_otp_wrdone()\n");
+	return -ENOEXEC;
+}
+
+static int req_otp_read_mode(void)
+{
+
+	write_word(OTP_RWSBMODE_ADDR, OTP_READ_MODE);
+
+	return poll_otp_ready();
+}
+
+
+static int req_otp_byte_write_mode(void)
+{
+
+	write_word(OTP_RWSBMODE_ADDR, OTP_BYTE_WRITE_MODE);
+
+	return poll_otp_ready();
+}
+
+static unsigned int read_otp_location_disp(unsigned int offset, unsigned int *read_val)
+{
+
+	int err;
+
+	write_word(OTP_RDENABLE_ADDR,  offset);
+	err = poll_otp_read_valid();
+	if (err) {
+		LOG_ERR("OTP read failed\n");
+		return err;
+	}
+	read_word(OTP_READREG_ADDR, read_val);
+
+	return 0;
+
+}
+
+static int  write_otp_location(unsigned int otp_location_offset, unsigned int otp_data)
+{
+	write_word(OTP_WRENABLE_ADDR,  otp_location_offset);
+
+	write_word(OTP_WRITEREG_ADDR,  otp_data);
+
+	return poll_otp_wrdone();
+}
+
+
+static int otp_rd_voltage_1V8(void)
+{
+	int err;
+
+	err = req_otp_standby_mode();
+
+	if (err) {
+		LOG_ERR("error in %s\n", __func__);
+		return err;
+	}
+	write_word(OTP_VOLTCTRL_ADDR, OTP_VOLTCTRL_1V8);
+	return 0;
+}
+
+
+int write_otp_memory(unsigned int otp_addr, unsigned int *write_val)
+{
+	int err;
+	int	mask_val;
+
+	err = poll_otp_ready();
+	if (err) {
+		LOG_ERR("err in otp ready poll\n");
+		return err;
+	}
+
+	set_otp_timing_reg_40mhz();
+
+	err = otp_wr_voltage_2V5();
+
+	if (err) {
+		LOG_ERR("error in write_voltage 2V5\n");
+		goto _exit_otp_write;
+	}
+
+	err = req_otp_byte_write_mode();
+	if (err) {
+		LOG_ERR("error in OTP byte write mode\n");
+		goto _exit_otp_write;
+	}
+
+	switch (otp_addr) {
+
+	case REGION_PROTECT:
+		write_otp_location(REGION_PROTECT, write_val[0]);
+		write_otp_location(REGION_PROTECT+1, write_val[0]);
+		write_otp_location(REGION_PROTECT+2, write_val[0]);
+		write_otp_location(REGION_PROTECT+3, write_val[0]);
+
+		LOG_INF("Written REGION_PROTECT0 (0x%x) : 0x%04x\n",
+						(REGION_PROTECT<<2), write_val[0]);
+		LOG_INF("Written REGION_PROTECT1 (0x%x) : 0x%04x\n",
+						(REGION_PROTECT+1)<<2, write_val[0]);
+		LOG_INF("Written REGION_PROTECT2 (0x%x) : 0x%04x\n",
+						(REGION_PROTECT+2)<<2, write_val[0]);
+		LOG_INF("Written REGION_PROTECT3 (0x%x) : 0x%04x\n",
+						(REGION_PROTECT+3)<<2, write_val[0]);
+		break;
+
+	case QSPI_KEY:
+		mask_val = QSPI_KEY_FLAG_MASK;
```

Comment:
this case block can be done in a loop/s

Reply:
This is a tested code that needs to go for customer sampling (needs to get merged today). Look into this for then next revision